Tests/src/artofillusion/SceneTest.java

```diff
@@ -10,32 +10,768 @@
+    public void testAddImageFromTexture() throws InterruptedException
+    {
+      int SIZE = 50;
+
+      ImageMapTexture tex = new ImageMapTexture();
+      BufferedImage im = new BufferedImage(SIZE, SIZE, BufferedImage.TYPE_INT_ARGB);
+      ImageMap map = new MIPMappedImage(im);
+      ImageOrColor ioc = new ImageOrColor(new RGBColor(), map);
+      tex.diffuseColor = ioc;
+      scene.addTexture(tex);
+      
+      assertEquals(1, scene.getNumImages());
+      assertEquals(0, scene.indexOf(map));
+              
+      assertEquals(map, scene.getImage(0));
+      
+    }
+    
+  /**
+   * Test adds new Texture to scene. Texture uses some image already added to scene.
+   * Test fails as image used by texture not added to scene implicitly
+   * @throws InterruptedException
+   */
+  @Test
+    public void testAddImageBeforeTexture() throws InterruptedException
+    {
+      int SIZE = 50;
+
+      ImageMapTexture tex = new ImageMapTexture();
+      BufferedImage im = new BufferedImage(SIZE, SIZE, BufferedImage.TYPE_INT_ARGB);
+      ImageMap map = new MIPMappedImage(im);
+      ImageOrColor ioc = new ImageOrColor(new RGBColor(), map);
+      tex.diffuseColor = ioc;
+      scene.addImage(map);
+      scene.addTexture(tex);
+      
+      assertEquals(1, scene.getNumImages());
+      assertEquals(0, scene.indexOf(map));
+              
+      assertEquals(map, scene.getImage(0));
+      
+    }    
+    
+  /**
+   * Test to remove image from scene
+   * @throws InterruptedException
+   */
+  @Test
+    public void testRemoveImage() throws InterruptedException
+    {
+      int SIZE = 50;
+
+      BufferedImage im = new BufferedImage(SIZE, SIZE, BufferedImage.TYPE_INT_ARGB);
+
+      ImageMap map = new MIPMappedImage(im);
+      scene.addImage(map);
+      boolean removed = scene.removeImage(0);
+      
+      assertTrue(removed);
+      assertEquals(0, scene.getNumImages());
+
+    }
+    
+  /**
+   * Test to remove image used through texture from scene
+   * Checks that image deletion is rejected
+   * @throws InterruptedException
+   */
+  @Test
+    public void testAttemptRemoveImageUsedInTexture() throws InterruptedException
+    {
+      int SIZE = 50;
+
+      ImageMapTexture tex = new ImageMapTexture();
+      BufferedImage im = new BufferedImage(SIZE, SIZE, BufferedImage.TYPE_INT_ARGB);
+      ImageMap map = new MIPMappedImage(im);
+      ImageOrColor ioc = new ImageOrColor(new RGBColor(), map);
+      tex.diffuseColor = ioc;
+      scene.addImage(map);
+      scene.addTexture(tex);
+      
+      boolean result = scene.removeImage(0);
+      assertFalse(result);
+      
+    }
+    
+  /**
+   * Test to remove image used through material from scene
+   * Checks that image deletion is rejected
+   * As no way to add image to Procedural Material as Image module programmatically test fails
+   * @throws InterruptedException
+   */
+  @Test
+    public void testAttemptRemoveImageUsedInMaterial() throws InterruptedException
```

Comment:
Generally, you won't use images in a `Material` at all. Not sure that this one is testable, or even makes sense.

Try adding using an `ImageModule,` perhaps in conjuction with a `ProceeduralTexture2d`. You're really trying to test if the image removal fails _when using the proceedural system_ anyway.

Reply:
Idea is to test that image used in Material prevent to be removed from scene, similar to image used in Texture which tested with testAttemptRemoveImageUsedInTexture(...). (This checks Scene's removeImage(...) method). 
But as no way to add image to Material programmatically I stub this test with fail(...)

Reply:
But you have not tested the general case of "Image used in `Texture`." You have tested the specific case of "Image used in `ImageMappedTexture`."

That is a perfectly good thing to test. The other thing to test is the case of "Image used in `Procedure`." This would have to be done with some extra work, perhaps by generating a special Object3d class in the test case that implements `ProcedureOwner`.  This would cover both the `ProceduralTexture` and `ProceduralMaterial` cases, as its the behavior of the `ImageModule` itself that would be of concern.